Add generic interactive shell

shihab-dls opened the pull request:
fixes #140 

This PR adds an interactive shell, based on this [implementation](https://github.com/GDYendell/fastcs2/blob/main/src/fastcs2/control_system.py#L44), and adds transport contexts. The `wip` commits are attempts to add `dbl` `dbpf` and `dbgf` functionality to `PVA` as an example, but maybe only the first commit is required for this PR, and further PRs can add transport specific context.

GDYendell commented on src/fastcs/launch.py:
```diff
@@ -100,12 +104,43 @@ def run(self):
+        if self._transports:
+            context = {"controller": self._backend}
+            for transport in self._transports:
+                context.update(transport.context())
```
We should have a checked update that errors if a key already exists. Hopefully it will never come up, but if it does we need to know.

GDYendell commented on src/fastcs/transport/adapter.py:
```diff
@@ -22,3 +22,7 @@ def create_docs(self) -> None:
+
+    @abstractmethod
+    def context(self) -> dict[str, Any]:
+        return {}
```
Make this not abstract so transports don't have to implement it

GDYendell commented on src/fastcs/transport/adapter.py:
```diff
@@ -22,3 +22,7 @@ def create_docs(self) -> None:
         pass
+
+    @abstractmethod
+    def context(self) -> dict[str, Any]:
```
This should be a property because it is not a verb

GDYendell reviewed (changes requested):
Minor comments

GDYendell commented on src/fastcs/launch.py:
```diff
@@ -100,12 +104,43 @@ def run(self):
         coros = [self._backend.serve()]
-        coros.extend([transport.serve() for transport in self._transports])
+        if self._transports:
+            context = {"controller": self._backend}
```
```suggestion
            context = {"controller": self._controller}
```

shihab-dls commented on src/fastcs/launch.py:
```diff
@@ -100,12 +105,52 @@ def run(self):
+            context.update(transport.context)
+
+        if not os.getenv("NOT_INTERACTIVE"):
+            coros.append(self._interactive_shell(context))
```
Hmm, we could make a `TestShell` fixture that hangs, and inject this into `FastCS.__init__`, storing the Iterminal in `self._ip_shell`? Something like:
```python
@pytest.fixture
def test_shell():
    class TestShell:
        def __init__(self):
            self._stop = threading.Event()

        def mainloop(self):
            self._stop.wait()

        def exit(self):
            self._stop.set()

    return TestShell()
```
We could then patch this like:
```python
@pytest.fixture(autouse=True)
def patch_fastcs_init_with_test_shell(test_shell):
    def wrapped_init(self, *args, **kwargs):
        fastcs.launch.FastCS.__init__(self, *args, **kwargs)
        self._ip_shell = test_shell

    with patch.object(fastcs.launch.FastCS, "__init__", wrapped_init):
        yield
```
I'm not sure if this is what you were going for though? I guess we would then need to find specific fastcs events in any test that executes `fastcs.run()` (which currently is the doc snippets) we want to trigger `test_shell.exit()`, or we could add another fixture that does `threading.Timer(delay, test_shell.exit)` to exit a bit after yielding to the test, although an event driven approach is probably better.

shihab-dls commented:
I've commented out `fastcs.run()` lines that were in some doc snippets for now, and removed the env var. Re-requesting a review with the generic terminal and the amended test. I've also added a test for the duplicate context error for coverage; let me know if you would've tested it another way.

Codecov is still failing, but for reasons I currently think are "okay"...

GDYendell commented:
I think I am OK with leaving that run method with missing lines unless you can come up with a neat way to test that

shihab-dls commented:
> I think I am OK with leaving that run method with missing lines unless you can come up with a neat way to test that

Hit my head against it for a little while, and can't figure out a "neat" way to do it; I can open up a small issue for me to look at after, but for now we should probably merge this in (if you are happy with my other test)?
